### PR TITLE
Gardenlet self-generates TLS server certificates if necessary

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area usability
/kind enhancement
/priority normal

**What this PR does / why we need it**:
With #2309 we have introduced a HTTPS server to the gardenlet which requires to specify a TLS server certificate + key during deployment time. This is quite uncomfortable as it requires to pre-generate such TLS keys.
With this PR, the gardenlet is enabled to self-generate such certificates during start-up.

**Special notes for your reviewer**:
It's not implemented for the gardener-controller-manager as it requires more effort - the reason is that it comes with a `ValidatingWebhookConfiguration` which needs to know the CA certificate, and this resource is not created by the GCM itself but via the Gardener Helm chart.

**Release note**:

```improvement operator
NONE
```
(I'll adapt the release note of #2309 once this PR is merged)

### CODE REVIEW

File: pkg/utils/secrets/certificates.go [hunk 1]
```diff
@@ -355,3 +357,45 @@ func GenerateCertificateAuthorities(k8sClusterClient kubernetes.Interface, exist
...
+	caCertificate, err := caCertificateConfig.GenerateCertificate()
+	if err != nil {
+		return nil, "", "", err
+	}
+
+	certificateConfig := &CertificateSecretConfig{
+		Name:       name,
+		CommonName: name,
+		DNSNames:   dnsNames,
+		CertType:   ClientCert,
```
Comment (vpnachev) [inline pkg/utils/secrets/certificates.go]: Shouldn't it be server certificate type?
```suggestion
		CertType:   ServerCert,
```

File: pkg/utils/secrets/certificates.go [hunk 2]
```diff
@@ -355,3 +357,45 @@ func GenerateCertificateAuthorities(k8sClusterClient kubernetes.Interface, exist
 
 	return generatedSecrets, certificateAuthorities, nil
 }
+
+// SelfGenerateTLSServerCertificate generates a new CA certificate and signs a server certificate with it. It'll return
+// the certificate objects and store the bytes in a temporary directory whose paths will also be returned.
+func SelfGenerateTLSServerCertificate(name string, dnsNames []string) (*Certificate, string, string, error) {
+	certificateDir, err := ioutil.TempDir("", "self-generated-server-certificates")
```
Comment (vpnachev) [inline pkg/utils/secrets/certificates.go]:
```suggestion
	certificateDir, err := ioutil.TempDir("", "self-generated-server-certificates-")
```

For better readability , currently it is looking like `/tmp/self-generated-server-certificates394803357/` but `/tmp/self-generated-server-certificates-394803357/` is looking better.

File: pkg/utils/secrets/certificates.go [hunk 3]
```diff
@@ -355,3 +357,45 @@ func GenerateCertificateAuthorities(k8sClusterClient kubernetes.Interface, exist
...
+	}
+	certPath := filepath.Join(certificateDir, "tls.pem")
+	keyPath := filepath.Join(certificateDir, "tls-key.pem")
+
+	caCertificateConfig := &CertificateSecretConfig{
+		Name:       name,
+		CommonName: name,
+		CertType:   CACert,
+	}
+	caCertificate, err := caCertificateConfig.GenerateCertificate()
```
Comment (vpnachev) [inline pkg/utils/secrets/certificates.go:376]: Please, save CA files, too.

**Review by vpnachev — COMMENTED** [review]
Very nice!

Just some minor nits in the certificate generation.

**Comment by rfranzke** [comment]
Thanks!
/needs review
/cc @vpnachev

**Review by vpnachev — APPROVED** [review]
/lgtm

**Pull request merged**